spec: Faster __hash__

vbrunini opened the pull request:
Split out of #51518 based on reviewer comments @haampie

vbrunini committed:
spec: Cache result of __hash__ for non-concrete specs.

Signed-off-by: Victor Brunini <vebruni@sandia.gov>

vbrunini committed:
spec: Include fewer properties in hash function for non-concrete specs.

This gives similar performance to caching the hash in the prior commit
but should avoid the issues with Spec's being mutable brought up in review.

Signed-off-by: Victor Brunini <vebruni@sandia.gov>

vbrunini committed:
Fix style.

Signed-off-by: Victor Brunini <vebruni@sandia.gov>

vbrunini commented:
> Specs are mutable, this breaks the invariant `x == y => hash(x) == hash(y)`. A better approach is to make the hash function cheaper (at the possible cost of more collisions in dicts and sets)

Tangent to this PR, but I started looking at the concretize setup time for my development environment as well and it spends a large fraction of time formatting Spec's to strings repeatedly for the same Spec's. I hacked in a similar change to the initial pass of caching the hash from this PR, but now caching the result of `__str__`. It would have some similar issues with Spec mutability I think, but it reduced concretize setup time from ~30s to ~20s.

haampie commented:
I think something like this is sufficient?

```python
        if not self._dependencies:  # fast path for trivial abstract specs (followed 94% in my tests)
            return hash((
                self.name,  # cheap
                self.namespace,  # cheap
                self.versions,
                self.variants if self.variants.dict else None,  # avoids expensive function call
                self.architecture,
                self.abstract_hash,  # cheap
            ))

        return hash(lang.tuplify(self._cmp_iter))
```

vbrunini committed:
spec.py: Updated spec __hash__ improvement from reviewer comments.

Timing "spack-python -c 'import spack.repo; x = list(spack.repo.PATH.all_package_classes())'"
I see:
Before: 15.5s
After: 14.8s

Signed-off-by: Victor Brunini <vebruni@sandia.gov>

haampie commented on lib/spack/spack/spec.py:
```diff
@@ -5052,9 +5052,20 @@ def __hash__(self):
+        if not self._dependencies:
+            return hash(
+                (
+                    self.name,  # cheap
```
Can you remove these 4 comments?

vbrunini committed:
spec.py: Remove comments.

Signed-off-by: Victor Brunini <vebruni@sandia.gov>

haampie reviewed (approved):
Thanks, looks good to me.
